src/main/scala/pl/datart/csvtojson/service/TaskRunner.scala

```diff
@@ -0,0 +1,131 @@
+package pl.datart.csvtojson.service
+
+import akka.http.scaladsl.model.Uri
+import better.files.File
+import cats.effect._
+import cats.effect.kernel.Resource.ExitCase
+import cats.effect.std._
+import cats.syntax.all._
+import fs2._
+import fs2.data.csv._
+import fs2.io.file.Files
+import pl.datart.csvtojson.model._
+import spray.json.DefaultJsonProtocol._
+import spray.json._
+
+import java.net.URL
+import java.nio.file.Paths
+import java.nio.file.StandardOpenOption._
+import scala.concurrent.duration._
+
+trait TaskRunner[F[_]] {
+  def run(taskId: TaskId, uri: Uri): F[Unit]
+}
+
+@SuppressWarnings(
+  Array("org.wartremover.warts.Any", "org.wartremover.warts.ImplicitParameter", "org.wartremover.warts.Nothing")
+)
+class TaskRunnerImpl[F[_]](taskService: TaskService[F], semaphore: Semaphore[F])(implicit
+    async: Async[F]
+) extends TaskRunner[F] {
+  override def run(taskId: TaskId, uri: Uri): F[Unit] = {
+    getAndSave(taskId, uri)
+  }
+
+  private def getAndSave(taskId: TaskId, uri: Uri): F[Unit] = {
+    for {
+      _          <- semaphore.acquire
```

Comment:
I think `acquire/release` should have defined boundaries.  It is hard to follow the code where `release` is called from multiple places, and what is more important, it won't be released if e.g.  `taskService.getTask` fails. I would propose to use e.g. `bracket` here or at least `attempt` (won't work with cancellable effects)...

Reply:
Indeed, it could be rebalanced, which I did.